programs/uxd/src/instructions/mercurial/collect_profits_of_mercurial_vault_depository.rs

```diff
@@ -0,0 +1,273 @@
+        controller_version: ctx.accounts.controller.load()?.version,
+        depository_version: ctx.accounts.depository.load()?.version,
+        controller: ctx.accounts.controller.key(),
+        depository: ctx.accounts.depository.key(),
+        collateral_amount: profits_beneficiary_collateral_amount_increase,
+    });
+
+    // 11 - update accounting
+    let current_time_as_unix_timestamp = u64::try_from(Clock::get()?.unix_timestamp)
+        .ok()
+        .ok_or(UxdError::MathError)?;
+
+    ctx.accounts
+        .depository
+        .load_mut()?
+        .update_onchain_accounting_following_profits_collection(
+            profits_beneficiary_collateral_amount_increase,
+            current_time_as_unix_timestamp,
+        )?;
+
+    ctx.accounts
+        .controller
+        .load_mut()?
+        .update_onchain_accounting_following_profit_collection(
+            profits_beneficiary_collateral_amount_increase,
+        )?;
+
+    Ok(())
+}
+
+// Into functions
+impl<'info> CollectProfitsOfMercurialVaultDepository<'info> {
+    pub fn into_withdraw_collateral_from_mercurial_vault_context(
+        &self,
+    ) -> CpiContext<
+        '_,
+        '_,
+        '_,
+        'info,
+        mercurial_vault::cpi::accounts::DepositWithdrawLiquidity<'info>,
+    > {
+        let cpi_accounts = mercurial_vault::cpi::accounts::DepositWithdrawLiquidity {
+            vault: self.mercurial_vault.to_account_info(),
+            token_vault: self.mercurial_vault_collateral_token_safe.to_account_info(),
+            lp_mint: self.mercurial_vault_lp_mint.to_account_info(),
+            user_token: self.profits_beneficiary_collateral.to_account_info(),
+            user_lp: self.depository_lp_token_vault.to_account_info(),
+            user: self.depository.to_account_info(),
+            token_program: self.token_program.to_account_info(),
+        };
+        let cpi_program = self.mercurial_vault_program.to_account_info();
+        CpiContext::new(cpi_program, cpi_accounts)
+    }
+}
+
+// Calculation/Check functions
+impl<'info> CollectProfitsOfMercurialVaultDepository<'info> {
+    fn calculate_lp_token_amount_to_match_collectable_profits_value(
+        &self,
+        target_value: u64,
+    ) -> Result<u64> {
+        let current_time = u64::try_from(Clock::get()?.unix_timestamp)
+            .ok()
+            .ok_or(UxdError::MathError)?;
+
+        // Because it's u64 type, we will never withdraw too much due to precision loss, but withdraw less.
+        // We withdraw less interests and fee due to precision loss and that's ok
+        Ok(self
+            .mercurial_vault
+            .get_unmint_amount(
+                current_time,
+                target_value,
+                self.mercurial_vault_lp_mint.supply,
+            )
+            .ok_or(UxdError::MathError)?)
+    }
+
+    pub fn calculate_collectable_profits_value(&self) -> Result<u64> {
+        let owned_lp_tokens_value: u128 =
```

Comment:
micro-nit: i think you can keep the whole computation in u64 here since there is no way to go above the limits? this would avoid back and forth casting

Reply:
`redeemable_amount_under_management` is u128, the conversion is necessary for the computation.

Reply:
true true, but "redeemable_amount_under_management" represents a "token.amount" which is capped at u64 in solana

Reply:
<img width="658" alt="Screenshot 2023-01-30 at 10 54 22 PM" src="https://user-images.githubusercontent.com/13584944/215514470-1b02bb86-01d9-4cd4-a9d4-31d31c8817d5.png">
yes can't mint over it, start wondering why are we having all these in u128 lol

Reply:
put the casting on `redeemable_amount_under_management` to address this